ModelUploader/Program.cs

```diff
@@ -117,6 +122,173 @@ static void Main(string[] args)
+                        break;
+                    case 400:
+                        // Model could not be uploaded because of a dependency 
+
+                        // first inspect Extends Section
+                        exitProcess = ProcessExtendsSection(file, dtdl);
+                        if (exitProcess) return true;
+
+                        exitProcess = ProcessPropertiesSection(file, dtdl);
+                        if (exitProcess) return true;
+
+                        // now try the original file back again
+                        exitProcess = UploadModel(file);
+
+                        break;
+                    default:
+                        break;
+                }
+            }
+            return exitProcess;
+        }
+
+        private static bool ProcessExtendsSection(string file, string dtdl)
+        {
+            bool exitProcess = false;
+            MatchCollection extendsSection = Regex.Matches(dtdl, "extends[\" :[\r\na-zA-Z0-9;_,]*]", RegexOptions.Singleline);
+            if (extendsSection.Count > 0)
+            {
+                string extendsSectionFull = extendsSection[0].Value;
+                MatchCollection dtmiExtends = Regex.Matches(extendsSectionFull, "dtmi:[ :[a-zA-Z0-9;_]*", RegexOptions.Singleline);
+
+                foreach (Match dtmiExtendsMatch in dtmiExtends)
+                {
+                    // find this model
+                    try
+                    {
+                        Response<ModelData> res = client.GetModel(dtmiExtendsMatch.Value);
+                        // model found! keep going
+                    }
+                    catch (RequestFailedException e)
+                    {
+                        // Model Not Found - find it in the directory and call Upload Model
+                        if (e.Status == 404)
+                        {
+                            string missingInterface = dtmiExtendsMatch.Value;
+                            int missingPartIndex = missingInterface.LastIndexOf(":");
+                            missingInterface = missingInterface.Substring(missingPartIndex + 1).Replace(";1", "");
+                            missingInterface = missingInterface.Replace(";", "");
+
+                            string[] missingFile = Directory.GetFiles(modelPath, missingInterface + ".json*", SearchOption.AllDirectories);
+                            if (missingFile.Count<string>() == 0)
+                            {
+                                // no file found, perhaps the definition of the schema is contained within the current file, lets try
+                                exitProcess = UploadModel(file);
+                                if (exitProcess == true)
+                                {
+                                    Log.Error($"Could not find a definition for Interace {" + missingInterface + "}");
+                                }
+                            }
+                            else {  
```

Comment:
What if missingFile.Count() > 1?

Reply:
This refers to the if clause starting on line 199. Don't know if that'll ever happen in the real world, but I guess it might, and if so, a warning or error log might be warranted?

Reply:
Added a warning